NiL.JS/BaseLibrary/Console.cs

```diff
@@ -1,68 +1,395 @@
+        {
+            LogMessage(LogLevel.Log, Tools.JSValueToObjectString(args[0]));
+            return JSValue.undefined;
+        }
+
+        //public JSValue dirxml(JSValue thisBind, Arguments args)
+        //{
+        //    return JSValue.undefined;
+        //}
+
+
+        public JSValue group(JSValue thisBind, Arguments args)
+        {
+            string label = Tools.FormatArgs(args) ?? "null";
+            if (label == "")
+                label = "console.group";
+
+            if (_groups.Count > 0)
+            {
+                string _temp = _groups[_groups.Count - 1];
+                _groups.RemoveAt(_groups.Count - 1);
+                LogMessage(LogLevel.Info, System.String.Format(_indentGroupBegin, label));
+                _groups.Add(_temp);
+            }
+            else
+                LogMessage(LogLevel.Info, System.String.Format(_indentGroupBegin0, label));
+            
+            _groups.Add(label);
+
+            return JSValue.undefined;
+        }
+
+        public JSValue groupCollapsed(JSValue thisBind, Arguments args)
+        {
+            group(thisBind, args);
+            return JSValue.undefined;
+        }
+
+        public JSValue groupEnd(JSValue thisBind, Arguments args)
+        {
+            if (_groups.Count == 0)
+                _groups.RemoveAt(_groups.Count - 1);
+            
+            return JSValue.undefined;
+        }
+
+
+        public JSValue time(JSValue thisBind, Arguments args)
+        {
+            string label = "";
+            if (args.length > 0)
+                label = (args[0] ?? "null").ToString();
+
+            if (_timers.ContainsKey(label))
+                _timers[label].Restart();
+            else
+                _timers.Add(label, Stopwatch.StartNew());
+
+            return JSValue.undefined;
+        }
+
+        public JSValue timeEnd(JSValue thisBind, Arguments args)
+        {
+            string label = "";
+            if (args.length > 0)
+                label = (args[0] ?? "null").ToString();
 
-            if (!JSObject.@is(f.Call(null), sample))
+            double elapsed = 0.0;
+            if (_timers.ContainsKey(label))
             {
-                var message = f.ToString();
-                message = message.Substring(message.IndexOf("=>") + 2).Trim();
-                System.Console.Error.WriteLine(message + " not equals " + sample);
+                _timers[label].Stop();
+                elapsed = (double)_timers[label].ElapsedTicks / Stopwatch.Frequency * 1000.0;
+                _timers.Remove(label);
             }
+            
+            if (label != "")
+                label += ": ";
+            LogMessage(LogLevel.Info, label + Tools.DoubleToString(System.Math.Round(elapsed, 10)) + "ms");
+
+            return JSValue.undefined;
         }
 
-        public static JSValue error(Arguments args)
+        
+
+        internal static List<console> _consoles;
```

Comment:
We really need to store all instances in static field? I think, this solution can bring to memory leak.